### PR TITLE
Exchange rates from ipfs

### LINKED ISSUE #121: implement Base Currency setting
Below you will find a list of all controller sent he app where user sees a fiat-currency equivalent of crypto amount.

- [ ] send controller
- [ ] transactions tab
- [ ] transaction info controller
- [ ] transaction status controller

Note that these amounts needs to be requested from the permanent IPFS record for that specific transaction.

@nk-the-crazy currently works on the implementation. followup up with him on the status before you begin working on that ticket. Once implemented we will be able to retrieve things like exchange rate and other transaction data for any blockchain in close tp real-time.

### PR DESCRIPTION
**Pull request by rafaelekol**
Close #121

### CODE REVIEW

File: app/src/main/java/bitcoin/wallet/modules/fulltransactioninfo/FullTransactionInfoInteractor.kt [hunk 1]
```diff
@@ -29,19 +30,33 @@ class FullTransactionInfoInteractor(private val adapter: IAdapter?, private val
...
             delegate?.didGetTransactionInfo(viewItem)
         }
     }
 
-    private fun getTransactionRecordViewItem(record: TransactionRecordNew, adapter: IAdapter): TransactionRecordViewItem {
-        val rates = exchangeRateManager.exchangeRates
-        val convertedValue = rates[adapter.coin.code]?.let { it * record.amount }
+    private fun fetchExchangeRate(transaction: TransactionRecord) {
+        transaction.timestamp?.let { timestamp ->
+            ExchangeRateManager.getRate(coinCode = transaction.coinCode, currency = DollarCurrency().code, timestamp = timestamp)
```
Comment (omurovch) [inline app/src/main/java/bitcoin/wallet/modules/fulltransactioninfo/FullTransactionInfoInteractor.kt]: Why manager class is used directly inside interactor? I'd suggest to use pass exchangeRateManager when initializing, and through interface.

File: app/src/main/java/bitcoin/wallet/modules/transactions/TransactionsInteractor.kt [hunk 2]
```diff
@@ -60,10 +60,40 @@ class TransactionsInteractor(private val adapterManager: AdapterManager, private
...
 
-        delegate?.didRetrieveItems(items)
+        items.sortByDescending { it.date }
+
+        disposables.add(Flowable.zip(flowableList, Arrays::asList)
+                .subscribeOn(Schedulers.io())
+                .unsubscribeOn(Schedulers.io())
+                .observeOn(io.reactivex.android.schedulers.AndroidSchedulers.mainThread())
+                .subscribe { resultRates ->
+                    val ratesMap = mutableMapOf<String, Double>()
```
Comment (abdrasulov) [inline app/src/main/java/bitcoin/wallet/modules/transactions/TransactionsInteractor.kt]: Please check if there is a shorter way for converting list of pairs to map

File: app/src/main/java/bitcoin/wallet/modules/wallet/WalletInteractor.kt [hunk 3]
```diff
@@ -14,7 +14,7 @@ class WalletInteractor(private val adapterManager: AdapterManager, private val e
     private var disposables: CompositeDisposable = CompositeDisposable()
 
     override fun notifyWalletBalances() {
-        adapterManager.subject.subscribe {
+        val disposable = adapterManager.subject.subscribe {
```
Comment (abdrasulov) [inline app/src/main/java/bitcoin/wallet/modules/wallet/WalletInteractor.kt]: The same as for the previous local disposable variable

File: app/src/main/java/bitcoin/wallet/modules/wallet/WalletInteractor.kt [hunk 4]
```diff
@@ -32,7 +32,7 @@ class WalletInteractor(private val adapterManager: AdapterManager, private val e
             progresses[adapter.id] = adapter.progressSubject
         }
 
-        val exchangeRates = ExchangeRateManager.exchangeRates
+        val exchangeRates = exchangeRateManager.exchangeRates
```
Comment (abdrasulov) [inline app/src/main/java/bitcoin/wallet/modules/wallet/WalletInteractor.kt]: Let's inline val exchangeRates

File: app/src/main/java/bitcoin/wallet/modules/transactions/TransactionsInteractor.kt (hunk 2 above)
Comment (rafaelekol) [reply app/src/main/java/bitcoin/wallet/modules/transactions/TransactionsInteractor.kt]: I couldn't find shorter version, maybe you can help me

File: app/src/main/java/bitcoin/wallet/core/Interfaces.kt [hunk 5]
```diff
@@ -18,8 +18,14 @@ interface IRandomProvider {
...
 
 interface INetworkManager {
-    fun getJwtToken(identity: String, pubKeys: Map<Int, String>): Observable<String>
-    fun getExchangeRates(): Observable<Map<String, Double>>
+    fun getLatestRate(coinCode:String, currency: String): Flowable<Double>
+    fun getRate(coinCode:String, currency: String, year: Int, month: String, day: String, hour: String, minute: String): Flowable<Double>
+    fun getRateByDay(coinCode:String, currency: String, year: Int, month: String, day: String): Flowable<Double>
+}
+
+interface IExchangeRateManager {
```
[the remark this answers is not in the document]
Comment (rafaelekol) [reply app/src/main/java/bitcoin/wallet/core/Interfaces.kt]: We need to use Interface in order to mock the Kotlin Object

[the remark this answers is not in the document]
Comment (abdrasulov) [reply app/src/main/java/bitcoin/wallet/core/Interfaces.kt]: What if we mock object? As far as I know Mockito allows to mock Kotlin object too.

**Pull request merged**